### PR TITLE
refactor: Align orientation and rotation terms

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
UiAutomator only operates the term of `Rotation`, which reflects the current rotation value of the device display. However,  the spec does not actually require that zero degrees rotation value is a portrait orientation, which might be the case for some devices. In such case the value returned for GetOrientation command would not be the one, that is naturally expected. 

In this PR we try to workaround that issue by matching the actual orientation value retrieved from instrumentation context resources configuration with the actual rotation value if the `useResourcesForOrientationDetection` setting is set to `true` (by default it is `false` to avoid issues with current automation scripts).

### CODE REVIEW

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  17
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  3
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  16
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

File: app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                return ROTATION_270;
+            default:
+                throw new IllegalArgumentException(
+                        String.format("Rotation value of %s degrees is not supported. " +
+                                "Only 0, 90, 180 and 270 degrees could be translated into " +
+                                "a valid screen rotation", degrees));
+        }
+    }
+
+    public static ScreenRotation ofOrientation(String abbr) {
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java:56]: Codacy found an issue: [The method 'ofOrientation(String)' has an NPath complexity of 480, current threshold is 200](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  16
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  2
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

File: app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java:56]: Codacy found an issue: [The method 'ofOrientation(String)' has an NPath complexity of 480, current threshold is 200](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/handler/SetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  16
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

File: app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java (hunk 2 above)
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java:56]: Codacy found an issue: [The method 'ofOrientation(String)' has an NPath complexity of 480, current threshold is 200](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/handler/SetOrientation.java  3
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  17
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  3
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

File: app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java:55]: Codacy found an issue: [The method 'ofOrientation(String)' has an NPath complexity of 480, current threshold is 200](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  15
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  1
- app/src/main/java/io/appium/uiautomator2/handler/SetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

File: app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java:55]: Codacy found an issue: [The method 'ofOrientation(String)' has an NPath complexity of 480, current threshold is 200](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  11
- app/src/main/java/io/appium/uiautomator2/handler/SetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  1
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  12
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  1
- app/src/main/java/io/appium/uiautomator2/handler/SetOrientation.java  2
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
- app/src/main/java/io/appium/uiautomator2/handler/SetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  12
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

File: app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                            ScreenOrientation.LANDSCAPE.name(), ScreenOrientation.PORTRAIT.name()));
+        }
+        ScreenRotation currentRotation = current();
+
+        if (!((UseResourcesForOrientationDetection) USE_RESOURCES_FOR_ORIENTATION_DETECTION.getSetting()).getValue()) {
+            return desiredOrientation == ScreenOrientation.LANDSCAPE ? ROTATION_270 : ROTATION_0;
+        }
+
+        ScreenOrientation currentOrientation = ScreenOrientation.current();
+        if (desiredOrientation == currentOrientation) {
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java:83]: Codacy found an issue: [Use equals() to compare object references.](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/handler/SetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  9
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

**Comment by mykola-mokhnach** [comment]
Linking https://github.com/appium/appium/issues/11663

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/handler/GetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/ScreenRotation.java  7
- app/src/main/java/io/appium/uiautomator2/handler/SetOrientation.java  2
- app/src/main/java/io/appium/uiautomator2/model/settings/UseResourcesForOrientationDetection.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/io/appium/uiautomator2/model/ScreenOrientation.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6386395&bid=20711547)

**Review by KazuCocoa — APPROVED** [review]
seems reasonable

**Pull request merged**